code/ATMOSPHERICS/hvac/spaceheater.dm

```diff
@@ -349,19 +349,30 @@
 
 /obj/machinery/space_heater/campfire/process()
 	..()
-	var/turf/T = get_turf(src)
+	if(!on)
+		return
+	var/turf/simulated/T = get_turf(loc)
 	var/datum/gas_mixture/env = T.return_air()
 	var/list/comfyfire = list('sound/misc/comfyfire1.ogg','sound/misc/comfyfire2.ogg','sound/misc/comfyfire3.ogg',)
 	if(Floor(cell.charge/10) != lastcharge)
 		update_icon()
-	if(!(cell && cell.charge > 0) && nocell != 2 | env.molar_density(GAS_OXYGEN) < 5 / CELL_VOLUME)
-		new /obj/effect/decal/cleanable/campfire(get_turf(src))
-		qdel(src)
+	if((!(cell && cell.charge > 0) && nocell != 2) || !istype(T) || (env.molar_density(GAS_OXYGEN) < 5 / CELL_VOLUME))
+		extinguish()
 		return
 	lastcharge = Floor(cell.charge/10)
 	if(on)
 		playsound(src, pick(comfyfire), (cell.charge/250)*5, 1, -1,channel = 124)
 
+/obj/machinery/space_heater/campfire/extinguish()
```

Comment:
I'd make a new proc rather than reusing this one.

Reply:
thought it might be handy this way, bit of centralization in fire-related procs, if there's ever a proper rewrite for fire the coder won't have to scuffle through old code for thousands of different ways different objects and their procs handle fire.
it just seems more logical like this

Reply:
I might agree if stoves/fireplaces/whatever made use of any of the other firecode-related things, but they don't